### PR TITLE
add function getOSKernelVersion

### PR DESCRIPTION
**Pull request by Joeywzr**
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Returns String with OS Kernel version.

Detailed description / Documentation draft:
#29599

> By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

> If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Interpreters/getOSKernelVersion.cpp [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    
+String getOSKernelVersion()
+{
+    struct utsname os_kernel_info;
+    int buf = uname(&os_kernel_info);
+    if (buf < 0)
+    {
+        throw Exception(
+            "EFAULT buffer is not valid.",
+            ErrorCodes::SYSTEM_ERROR);
```
Comment (azat) [inline src/Interpreters/getOSKernelVersion.cpp:21]:
```suggestion
        throw throwFromErrno("uname failed", ErrorCodes::SYSTEM_ERROR);
```

File: src/Interpreters/getOSKernelVersion.cpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+namespace ErrorCodes
+{
+    extern const int SYSTEM_ERROR;
+}
+    
+String getOSKernelVersion()
+{
+    struct utsname os_kernel_info;
+    int buf = uname(&os_kernel_info);
+    if (buf < 0)
```
Comment (azat) [inline src/Interpreters/getOSKernelVersion.cpp:17]: `uname` returns error (if any), so either

```suggestion
    int err = uname(&os_kernel_info);
    if (err)
```

Or 

```suggestion
    if (uname(&os_kernel_info))
```

File: src/Interpreters/getOSKernelVersion.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
+#pragma once
+#if defined(OS_LINUX)
+
+#include <Common/typeid_cast.h>
+
+#include <string>
+#include <sys/utsname.h>
```
Comment (azat) [inline src/Interpreters/getOSKernelVersion.h:7]: This can be moved into .cpp

File: src/Interpreters/getOSKernelVersion.cpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        throw Exception(
+            "EFAULT buffer is not valid.",
+            ErrorCodes::SYSTEM_ERROR);
+    }
+    else
+    {
+        return String(os_kernel_info.sysname) + " " + String(os_kernel_info.release);
```
Comment (azat) [inline src/Interpreters/getOSKernelVersion.cpp:25]: But how it differs from `Poco::Environment::osName()` and `Poco::Environment::osVersion()`

https://github.com/ClickHouse/ClickHouse/blob/dab4c60596f4e6897c3d4d5fca1c6f470e2d16a2/programs/server/Server.cpp#L413-L416

File: src/Interpreters/getOSKernelVersion.h [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace DB
+{
+
+/// Returns String with OS Kernel version.
+/* To get name and information about current kernel.
+   For simplicity, the function can be implemented only for Linux. 
+*/
+    
+String getOSKernelVersion();
```
Comment (azat) [inline src/Interpreters/getOSKernelVersion.h:17]: BTW right now it is not used anywhere, so at least logging it's result during startup should be added I guess (although for now it does not differs from `Poco::Environment::OS*` that is already printed at startup, see previous comment)

File: src/Interpreters/getOSKernelVersion.cpp (hunk 4 above)
Comment (Joeywzr) [reply src/Interpreters/getOSKernelVersion.cpp:25]: Maybe it is designed to natively support it. I just found that there are a lot of libraries in the contrib that contain this function.

File: src/Interpreters/getOSKernelVersion.h (hunk 5 above)
Comment (Joeywzr) [reply src/Interpreters/getOSKernelVersion.h:17]: Makes sense. What do you think? @alexey-milovidov

Comment (Joeywzr) [reply src/Interpreters/getOSKernelVersion.h:17]: Makes sense. What do you think? @alexey-milovidov

**Comment by nikitamikhaylov** [comment]
I think that you have to implement clickhouse's function, not pure c++ function. Look at `functions/serverContants.cpp`

**Comment by kitaisreal** [comment]
@Joeywzr yes, task is to write clickhouse function that will return OSKernelVersion.

**Comment by Joeywzr** [comment]
@nikitamikhaylov @kitaisreal Thanks! I'll modify my code.

**Review by kitaisreal — APPROVED** [review]

**Review by kitaisreal — APPROVED** [review]

**Comment by azat** [comment]
Egh,  #29913 breaks it and it required rebase, so now it breaks build in master.

**Comment by azat** [comment]
Fixed in #29931

**Pull request merged**